zerver/migrations/0673_rename_general_chat_to_empty_string_topic.py

```diff
@@ -0,0 +1,114 @@
+from typing import Any
+
+from django.db import connection, migrations, models, transaction
+from django.db.backends.base.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+from psycopg2.sql import SQL, Identifier
+
+
+def update_edit_history(message_model: type[Any], realm_id: int) -> None:
+    BATCH_SIZE = 10000
+    lower_id_bound = 0
+
+    max_id = message_model.objects.filter(realm_id=realm_id).aggregate(models.Max("id"))["id__max"]
+    if max_id is None:
+        return
+
+    while lower_id_bound < max_id:
+        upper_id_bound = min(lower_id_bound + BATCH_SIZE, max_id)
+        with connection.cursor() as cursor:
+            query = SQL(
+                """
+                UPDATE {table_name}
+                SET edit_history = (
+                    SELECT JSONB_AGG(
+                        elem
+                        ||
+                        (CASE
+                            WHEN elem ? 'prev_topic' AND elem->>'prev_topic' = 'test general chat'
+                            THEN '{{"prev_topic": ""}}'::jsonb
+                            ELSE '{{}}'::jsonb
+                        END)
+                        ||
+                        (CASE
+                            WHEN elem ? 'topic' AND elem->>'topic' = 'test general chat'
+                            THEN '{{"topic": ""}}'::jsonb
+                            ELSE '{{}}'::jsonb
+                        END)
+                    )::text
+                    FROM JSONB_ARRAY_ELEMENTS(edit_history::jsonb) AS elem
+                )
+                WHERE realm_id = %(realm_id)s
+                AND id > %(lower_id_bound)s AND id <= %(upper_id_bound)s
+                AND (
+                    edit_history::jsonb @> '[{{"prev_topic": "test general chat"}}]' OR
+                    edit_history::jsonb @> '[{{"topic": "test general chat"}}]'
+                );
+            """
+            ).format(table_name=Identifier(message_model._meta.db_table))
+            cursor.execute(
+                query,
+                {
+                    "realm_id": realm_id,
+                    "lower_id_bound": lower_id_bound,
+                    "upper_id_bound": upper_id_bound,
+                },
+            )
+
+        print(f"Processed {upper_id_bound} / {max_id}")
+        lower_id_bound += BATCH_SIZE
```

Comment:
This loop model is not a good fit for systems like Zulip Cloud; doing one iteration through every message ID for every realm is a very big loop. I think what you want to do is instead have step 1 be to get the set of channels that contain such a topic globally across all realms (we've previously checked they are rare and this will be fast), and then do one non-paginated query per channel.

Reply:
[#backend > general chat - migration discussion @ 💬](https://chat.zulip.org/#narrow/channel/3-backend/topic/general.20chat.20-.20migration.20discussion/near/2103759)